src/creatures/players/cast/cast_viewer.cpp

```diff
@@ -0,0 +1,1533 @@
+		for (const auto &[client, clientInfo] : m_viewers) {
+			if (asLowerCaseString(clientInfo.first) != *it) {
+				continue;
+			}
+
+			m_bans[*it] = client->getIP();
+			client->sendSessionEndInformation(SESSION_END_LOGOUT);
+		}
+	}
+}
+
+bool CastViewer::checkBannedIP(uint32_t ip) const {
+	for (const auto &it : m_bans) {
+		if (it.second == ip) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+std::shared_ptr<ProtocolGame> CastViewer::getCastOwner() const {
+	return m_owner;
+}
+
+void CastViewer::setCastOwner(std::shared_ptr<ProtocolGame> client) {
+	m_owner = client;
+}
+
+void CastViewer::resetCastOwner() {
+	m_owner.reset();
+}
+
+std::string CastViewer::getCastPassword() const {
+	return m_castPassword;
+}
+
+void CastViewer::setCastPassword(const std::string &value) {
+	m_castPassword = value;
+}
+
+bool CastViewer::isCastBroadcasting() const {
+	return m_castBroadcast;
+}
+
+void CastViewer::setCastBroadcast(bool value) {
+	m_castBroadcast = value;
+}
+
+std::string CastViewer::getCastBroadcastTimeString() const {
+	int64_t seconds = getCastBroadcastTime() / 1000;
+	uint16_t hour = floor(seconds / 60 / 60 % 24);
+	uint16_t minute = floor(seconds / 60 % 60);
+	uint16_t second = floor(seconds % 60);
+
+	return fmt::format("{} hours, {} minutes and {} seconds", hour, minute, second);
+}
+
+int64_t CastViewer::getCastBroadcastTime() const {
+	return OTSYS_TIME() - m_castBroadcastTime;
+}
+
+void CastViewer::setCastBroadcastTime(int64_t time) {
+	m_castBroadcastTime = time;
+}
+
+std::string CastViewer::getCastDescription() const {
+	return m_castDescription;
+}
+
+void CastViewer::setCastDescription(const std::string &desc) {
+	m_castDescription = desc;
+}
+
+uint32_t CastViewer::getViewerId(std::shared_ptr<ProtocolGame> client) const {
+	auto it = m_viewers.find(client);
+	if (it != m_viewers.end()) {
+		return it->second.second;
+	}
+	return 0;
+}
+
+// Inherited
+void CastViewer::insertCaster() {
+	if (m_owner) {
+		m_owner->insertCaster();
+	}
+}
+
+void CastViewer::removeCaster() {
+	if (m_owner) {
+		m_owner->removeCaster();
+	}
+}
+
+bool CastViewer::canSee(const Position &pos) const {
+	if (m_owner) {
+		return m_owner->canSee(pos);
+	}
+
+	return false;
+}
+
+uint32_t CastViewer::getIP() const {
+	if (m_owner) {
+		return m_owner->getIP();
+	}
+
+	return 0;
+}
+
+void CastViewer::sendStats() {
+	if (m_owner) {
+		m_owner->sendStats();
+
+		for (const auto &it : m_viewers) {
+			it.first->sendStats();
+		}
+	}
+}
+
+void CastViewer::sendPing() {
+	if (m_owner) {
+		m_owner->sendPing();
+
+		for (const auto &it : m_viewers) {
+			it.first->sendPing();
+		}
+	}
+}
+
+void CastViewer::logout(bool displayEffect, bool forceLogout) {
+	if (m_owner) {
+		m_owner->logout(displayEffect, forceLogout);
+	}
+}
+
+void CastViewer::sendAddContainerItem(uint8_t cid, uint16_t slot, std::shared_ptr<Item> item) {
+	if (m_owner) {
+		m_owner->sendAddContainerItem(cid, slot, item);
+
+		for (const auto &it : m_viewers) {
```

Comment:
Implemented here, please if you can review and give me feedback: b09fce73d

Reply:
It's hard to say, does the dispatcher provide strong ordering guarantees under `TaskGroup::GenericParallel`? I would imagine tile operations (add, remove, update, etc) causing client issues if they happened to execute out of turn.

Reply:
> It's hard to say, does the dispatcher provide strong ordering guarantees under `TaskGroup::GenericParallel`? I would imagine tile operations (add, remove, update, etc) causing client issues if they happened to execute out of turn.

Thank you for the feedback! I understand the concerns regarding potential ordering issues with TaskGroup::GenericParallel, especially with tile operations. I haven't fully analyzed the implications yet, but I will take a closer look into how the dispatcher manages the operation sequence to ensure it won’t cause issues for clients. Once I have a better understanding, I'll provide a more definitive response.

Reply:
@dudantas 
```c++
	// Make a copy of viewers in case they can be modified while being processed in another thread.
	threadPool.detach_task([viewers = m_viewers, func]() {
		for (const auto &[client, _] : viewers) {
			func(client);
		}
	});
	````

Reply:
> @dudantas
> 
> ```c++
> 	// Make a copy of viewers in case they can be modified while being processed in another thread.
> 	threadPool.detach_task([viewers = m_viewers, func]() {
> 		for (const auto &[client, _] : viewers) {
> 			func(client);
> 		}
> 	});
> ```

I tested this, but it kept giving "debug" on the client, probably due to the order of the elements.